### DISCUSSION TITLE
Morphia 2.2.10, Mongo driver 4.4.0, Save operation fails for could not extract shard key

### DISCUSSION THREAD

**Discussion by arivazhaganj** [MorphiaOrg/morphia#2424] [Q&A]
With reference to similar issue discussed/reported: https://github.com/MorphiaOrg/morphia/issues/1872

we are using Morphia 2.2.10 and Mongo driver 4.4.0.  (currently app runs with compatilibity version of 4.0 of Mongo DB)

During upgrading mongo DB from 4.0 to 4.2, getting could not extract shard Key error.

We have sharded collection and we are trying to Save the sharded entity. Verified the sharded Entity holds value for the Shard Key.

But during Save operation, getting could not extract shard Key error.

```
com.mongodb.MongoWriteException: Write operation error on server localhost:27017. Write error: WriteError{code=61, message='Failed to target upsert by query :: could not extract exact shard key', details={}}.
	at com.mongodb.client.internal.MongoCollectionImpl.executeSingleWriteRequest(MongoCollectionImpl.java:1018) ~[mongodb-driver-sync-4.4.0.jar:na]
	at com.mongodb.client.internal.MongoCollectionImpl.executeReplaceOne(MongoCollectionImpl.java:568) ~[mongodb-driver-sync-4.4.0.jar:na]
	at com.mongodb.client.internal.MongoCollectionImpl.replaceOne(MongoCollectionImpl.java:551) ~[mongodb-driver-sync-4.4.0.jar:na]
	at dev.morphia.DatastoreImpl.save(DatastoreImpl.java:512) ~[morphia-core-2.2.10.jar:na]
	at dev.morphia.DatastoreImpl.save(DatastoreImpl.java:397) ~[morphia-core-2.2.10.jar:na]

```

Any help or suggestions to fix this?

1. as per other discussions, passing in filter with shard key will fix it. Not sure how to pass filter criteria for a Save Operation.
2. Modifying Shard key is not an option, since we currently trying to move to Mongo DB 4.2 only. 
Please help and share suggestions @evanchooly.

**Comment by evanchooly** [comment]
Your entity should have a mapped shard key property set before saving.  Is that property missing or null?

**Reply by arivazhaganj** [reply]
> Your entity should have a mapped shard key property set before saving. Is that property missing or null?

@ShardKey annotation is available since 2.3 of Morphia right? we are currently on 2.2.10.. please let us know if any other configuration is needed here @evanchooly

**Reply by arivazhaganj** [reply]
![Screenshot from 2023-06-12 18-27-48](https://github.com/MorphiaOrg/morphia/assets/35175748/a8bbbd60-4d86-4378-8134-9cdb6a03c091)

The above screenshot show, save method applying filter only by _id.. so may be shard key (which is other than _id) might fail here?

Please suggest

**Comment by evanchooly** [comment]
Morphia would need that shard key, yes, so I would recommend upgrading to 2.3.

**Reply by arivazhaganj** [reply]
> Morphia would need that shard key, yes, so I would recommend upgrading to 2.3.

Thank you @evanchooly .. Not sure on possibility of 2.3 upgrade immediately.. will plan on this.. But is there any work around available with 2.2.10 or 2.2.x please? For an immediate fix

**Comment by evanchooly** [comment]
For 2.2.x, there isn't I'm afraid.  If you want sharding support, you'll need to upgrade.
